aptos-move/block-executor/src/executor.rs

```diff
@@ -934,6 +950,65 @@ where
                     runtime_environment,
                     final_results,
                 )?;
+
+                if !end_block {
+                    continue;
+                }
+
+                if let Some(block_id) =
+                    transaction_slice_metadata.append_state_checkpoint_to_block()
+                {
+                    let mut outputs = final_results.acquire();
+                    let last_non_retry = outputs
+                        .iter()
+                        .rposition(|t| !t.committed_status().is_retry());
+                    let has_reconfig = if let Some(idx) = last_non_retry {
+                        outputs[idx].has_new_epoch_event()
+                    } else {
+                        false
+                    };
+
+                    if !has_reconfig {
+                        let txn = Self::gen_block_epilogue(
+                            block_id,
+                            block,
+                            outputs.dereference(),
+                            shared_commit_state.acquire().get_block_end_info(),
+                        );
+                        outputs.dereference_mut().push(E::Output::skip_output()); // placeholder
+                        if Self::execute(
```

Comment:
cc: @gelash if there's better way to append a txn at the end of execution, this looks pretty ugly.